scripts/hud/comparisons.ts

```diff
@@ -1,140 +1,325 @@
+			}),
+			'',
+			{
+				class: 'hud-splits__time',
+				text: '{g:time:time}'
+			}
+		),
+		diff: $.CreatePanel(
+			'Label',
+			$.CreatePanel('Panel', this.panels.splits.diffs, '', {
+				class: 'hud-splits__cell hud-splits__cell--diff ' + (i === 0 ? 'hud-splits__cell--latest' : '')
+			}),
+			'',
+			{
+				class: 'hud-splits__diff diff',
+				text: '{s:diff_sign}{g:time:diff}'
+			}
+		)
+	}));
+
+	// - When timer progression changes:
+	//   - Generate progression split, set bottom-most split to it
+	//   - Push all other splits up by setting them to the previous split's properties
+	// - When timer state changes,
+	//   - To primed: Clear all splits
+	//   - To running: Clear all splits
+	//   - To disabled: Clear all splits
+	//   - To finished: Generate final split, set bottom-most split to it
+	// - When comparison changes:
+	//   - Update all splits to use new comparison
+	// - When timer replaced:
+	//   - Regenerate everything. During seeking we'll do this a *lot*, but this
+	//     design is very performant and can handle it. If we ever need to speed it
+	//     up, could add an event specifically for seeking.
 	constructor() {
-		$.RegisterForUnhandledEvent('ComparisonRunUpdated', () => this.onComparisonRunUpdated());
-		$.RegisterForUnhandledEvent('OnObservedTimerStateChange', () => this.updateComparisons());
-		$.RegisterForUnhandledEvent('OnObservedTimerCheckpointProgressed', () => this.updateComparisons());
-		$.RegisterForUnhandledEvent('OnObservedTimerReplaced', () => this.updateComparisons());
-	}
+		$.RegisterForUnhandledEvent('OnObservedTimerStateChange', () => {
+			const { state, runTime, segmentsCount, segmentCheckpointsCount } =
+				MomentumTimerAPI.GetObservedTimerStatus();
 
-	onComparisonRunUpdated() {
-		this.comparison = RunComparisonsAPI.GetComparisonRun();
+			if (state === Timer.TimerState.FINISHED && (segmentsCount > 1 || segmentCheckpointsCount > 1)) {
+				const splits = MomentumTimerAPI.GetObservedTimerRunSplits();
 
-		$.Msg(`Got comparison run ${this.comparison?.filePath ?? 'NONE'}`);
+				this.updateLatestSplit(
+					Timer.generateFinishSplit(
+						splits,
+						this.comparison?.runSplits ?? null,
+						runTime,
+						this.comparison?.runTime ?? 0,
+						segmentsCount,
+						segmentCheckpointsCount
+					)
+				);
+			} else {
+				this.clearSplits();
+			}
+		});
 
-		if (this.comparison) {
-			$.Msg(`jumps: ${this.comparison.runSplits.trackStats.jumps}`);
-		}
+		$.RegisterForUnhandledEvent('ComparisonRunUpdated', () => {
+			const { state } = MomentumTimerAPI.GetObservedTimerStatus();
+			// If timer just finished and we're not watching a replay, don't update - otherwise
+			// comparison will get set to the run you just finished when you PB which we
+			// obviously don't want to happen.
+			if (state === Timer.TimerState.FINISHED && this.controlledReplayID === null) {
+				return;
+			}
+
+			this.comparison = RunComparisonsAPI.GetComparisonRun();
```

Comment:
I saw a case where the splits HUD was comparing against what I'm pretty sure was an older comparison one than the one I had set (the center timer comparison HUD was showing the delta I expected to see). This might mean the game code is not calling `ComparisonRunUpdated` when it should be though so probably not a bug here. I didn't dig into it more because I was debugging something else at the time and ran out of energy/time to try to recreate it.

Reply:
I think this code was wrong (below), not setting comparison to latest.
```ts
if (state === Timer.TimerState.FINISHED && this.controlledReplayID === null) {
	return;
}

this.comparison = RunComparisonsAPI.GetComparisonRun();
this.recomputeComparisons();
```
should be
```ts
this.comparison = RunComparisonsAPI.GetComparisonRun();

if (!(state === Timer.TimerState.FINISHED && this.controlledReplayID === null)) {
	this.recomputeComparisons();
}
```

Reply:
Fixed in `995469b`